File: src/main/java/org/devcon/ticket/PublicIdentifierProof.java
package org.devcon.ticket;

import java.math.BigInteger;
import java.util.Arrays;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.math.ec.ECPoint;
import org.tokenscript.attestation.FullProofOfExponent;
import org.tokenscript.attestation.IdentifierAttestation.AttestationType;
import org.tokenscript.attestation.core.AttestationCrypto;
import org.tokenscript.attestation.core.ExceptionUtil;
import org.tokenscript.attestation.core.Verifiable;

public class PublicIdentifierProof implements Verifiable {
  private static final Logger logger = LogManager.getLogger(PublicIdentifierProof.class);

  private final byte[] referenceCommitment;
  private final String identifier;
  private final AttestationType type;
  private final FullProofOfExponent internalPok;

  public PublicIdentifierProof(AttestationCrypto crypto, byte[] commitment, String identifier, AttestationType type, BigInteger secret) {
    this.referenceCommitment = commitment;
    this.identifier = identifier;
    this.type = type;
    this.internalPok = crypto.computeAttestationProof(secret);
    constructorCheck();
  }

  public PublicIdentifierProof(byte[] commitment, String identifier,
      AttestationType type, FullProofOfExponent pok) {
    this.referenceCommitment = commitment;
    this.identifier = identifier;
    this.type = type;
    this.internalPok = pok;
    constructorCheck();
  }

  private void constructorCheck() {
    if (!verify()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Proof, commitment and email not consistent or not valid"));
    }
  }
  public FullProofOfExponent getInternalPok()  {
    return internalPok;
  }

  @Override
  public boolean verify() {
    if (!verifyCommitment()) {
      return false;
    }
    if (!AttestationCrypto.verifyFullProof(internalPok)) {
      return false;
    }
    return true;
  }

  private boolean verifyCommitment() {
    BigInteger hashedIdentifier = AttestationCrypto.mapToCurveMultiplier(type, identifier);
    ECPoint hiddenMessagePoint = AttestationCrypto.G.multiply(hashedIdentifier);
    byte[] expectedCommitment = internalPok.getRiddle().add(hiddenMessagePoint).getEncoded(false);
    if (!Arrays.equals(expectedCommitment, referenceCommitment)) {
      return false;
    }
    return true;
  }

}

File: src/main/javascript/crypto/src/libs/PublicIdentifierProof.ts
import { AttestationCrypto, Pedestren_G } from "./AttestationCrypto";
import { FullProofOfExponent } from "./FullProofOfExponent";
import { Point } from "./Point";
import { uint8tohex } from "./utils";
import { Verifiable } from "./Verifiable";

export class PublicIdentifierProof implements Verifiable {
    private referenceCommitment: Uint8Array;
    private identifier: string;
    private type: number ;
    private internalPok: FullProofOfExponent;
    private crypto: AttestationCrypto;


    constructor() {
        this.crypto = new AttestationCrypto();
    }

    static fromSecret(commitment: Uint8Array, identifier: string, type: number, secret: bigint) {
        let pok = new AttestationCrypto().computeAttestationProof(secret);
        return PublicIdentifierProof.fromPOK(commitment, identifier, type, pok);
    }

    static fromPOK(commitment: Uint8Array, identifier: string,
    type: number, pok: FullProofOfExponent):PublicIdentifierProof {
        let me = new this();
        me.referenceCommitment = commitment;
        me.identifier = identifier;
        me.type = type;
        me.internalPok = pok;
        me.constructorCheck();
        return me;
    }

    private constructorCheck() {
        if (!this.verify()) {
            throw new Error("Proof, commitment and email not consistent or not valid");
        }
    }
    public getInternalPok(): FullProofOfExponent  {
        return this.internalPok;
    }

    public verify():boolean {
        if (!this.verifyCommitment()) {
            return false;
        }
        if (!new AttestationCrypto().verifyFullProof(this.internalPok)) {
            return false;
        }
        return true;
    }

    private verifyCommitment():boolean {
        let hashedIdentifier:bigint = this.crypto.mapToCurveMultiplier(this.type, this.identifier);
        let hiddenMessagePoint:Point  = Pedestren_G.multiplyDA(hashedIdentifier);
        let expectedCommitment: Uint8Array = this.internalPok.getRiddle().add(hiddenMessagePoint).getEncoded();
        if (uint8tohex(expectedCommitment) != uint8tohex(this.referenceCommitment)) {
            return false;
        }
        return true;
    }
}